### ISSUE TITLE
"None" check false-positive stemming from re.match.group(n)

### ISSUE THREAD

**Issue by wade-arista** [microsoft/pyright#7842]
**Describe the bug**
When assigning the output of `m.group(<N>)` to an optional `str`, even after an explicit `assert`, the variable is still `str | None`. This did not happen with a simple class and member function that returns `str | None`, but seems more related to `re.Match.group` in some way,

**Code or Screenshots**

```python
import re

def func(candidate: str) -> str:
    val: str | None = None

    m = re.match(r"^(a)?.*(b)$", candidate)
    if m:
        part2 = m.group(2)
        assert part2
        val = part2.strip()
        return val  # Expression of type "str | None" is incompatible with return type "str"

    return ""
```

```
$ pyright check.py
check.py
  check.py:11:16 - error: Expression of type "str | None" is incompatible with return type "str"
    Type "str | None" is incompatible with type "str"
      "None" is incompatible with "str" (reportReturnType)
1 error, 0 warnings, 0 informations 
```

```
$ pyright --version
pyright 1.1.361
$ python --version
Python 3.10.12
$ 
```

**Counter example**
```python
class Helper:
    def maybe(self, candidate: str) -> str | None:
        return "hi" if len(candidate) < 10 else None

def func(candidate: str) -> str:
    val: str | None = None

    m = Helper() if candidate else None
    if m:
        part2 = m.maybe(candidate)
        assert part2
        val = part2.strip()
        return val  # No complaints

    return ""
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

The issue here stems from the fact that `m.group(2)` evaluates to the type `str | Any`. The `Any` comes from the return type of one of the overloads for `Match.group`. This expression is assigned to variable `part2`. The statement `assert part2` is not able to further narrow the type `str | Any`. The expression `part2.strip()` evaluates to `str | Any`. This is assigned to the variable `var` which has a declared type of `str | None`. The type `str | None` cannot be narrowed upon assignment in this case, so the type of `val` remains `str | None`.

In your counter example, the expression `m.maybe(candidate)` evaluates to `str | None` rather than `str | Any`.

**Comment by wade-arista** [comment]
> In your counter example, the expression `m.maybe(candidate)` evaluates to `str | None` rather than `str | Any`.

Many thanks, Eric, and sorry for the noise. When I had inserted some `reveal_type` statements, I think I read `str | Any` as `str | None`, since it now very obvsiously shows up in the way you've described.
